File: android/app/src/main/java/com/example/clawix/android/core/BridgeFrameValidation.kt
package com.example.clawix.android.core

import kotlinx.serialization.json.JsonObject

internal object BridgeFrameValidation {
    private val topLevelKeys = setOf("schemaVersion", "type")

    private val allowedPayloadKeys: Map<String, Set<String>> = mapOf(
        "auth" to setOf("token", "deviceName", "clientKind", "clientId", "installationId", "deviceId"),
        "listSessions" to emptySet(),
        "openSession" to setOf("sessionId", "limit"),
        "loadOlderMessages" to setOf("sessionId", "beforeMessageId", "limit"),
        "sendMessage" to setOf("sessionId", "text", "attachments"),
        "newSession" to setOf("sessionId", "text", "attachments"),
        "interruptTurn" to setOf("sessionId"),
        "authOk" to setOf("hostDisplayName"),
        "authFailed" to setOf("reason"),
        "versionMismatch" to setOf("serverVersion"),
        "sessionsSnapshot" to setOf("sessions"),
        "sessionUpdated" to setOf("session"),
        "messagesSnapshot" to setOf("sessionId", "messages", "hasMore"),
        "messagesPage" to setOf("sessionId", "messages", "hasMore"),
        "messageAppended" to setOf("sessionId", "message"),
        "messageStreaming" to setOf("sessionId", "messageId", "content", "reasoningText", "finished"),
        "errorEvent" to setOf("code", "message"),
        "editPrompt" to setOf("sessionId", "messageId", "text"),
        "archiveSession" to setOf("sessionId"),
        "unarchiveSession" to setOf("sessionId"),
        "pinSession" to setOf("sessionId"),
        "unpinSession" to setOf("sessionId"),
        "renameSession" to setOf("sessionId", "title"),
        "pairingStart" to emptySet(),
        "listProjects" to emptySet(),
        "readFile" to setOf("path"),
        "pairingPayload" to setOf("qrJson", "token", "shortCode"),
        "projectsSnapshot" to setOf("projects"),
        "fileSnapshot" to setOf("path", "content", "isMarkdown", "error"),
        "transcribeAudio" to setOf("requestId", "audioBase64", "mimeType", "language"),
        "transcriptionResult" to setOf("requestId", "text", "errorMessage"),
        "requestAudio" to setOf("audioId"),
        "audioSnapshot" to setOf("audioId", "audioBase64", "mimeType", "errorMessage"),
        "requestGeneratedImage" to setOf("path"),
        "generatedImageSnapshot" to setOf("path", "dataBase64", "mimeType", "errorMessage"),
        "requestRolloutAttachment" to setOf("attachmentId"),
        "rolloutAttachmentSnapshot" to setOf("attachmentId", "dataBase64", "mimeType", "errorMessage"),
        "bridgeState" to setOf("state", "chatCount", "message"),
        "requestRateLimits" to emptySet(),
        "rateLimitsSnapshot" to setOf("rateLimits", "rateLimitsByLimitId"),
        "rateLimitsUpdated" to setOf("rateLimits", "rateLimitsByLimitId"),
        "requestClawJSServiceStatuses" to emptySet(),
        "clawJSServiceStatusesSnapshot" to setOf("services"),
        "clawJSServiceStatusUpdated" to setOf("service"),
        "audioRegister" to setOf("requestId", "request"),
        "audioAttachTranscript" to setOf("requestId", "audioId", "transcript"),
        "audioGet" to setOf("requestId", "audioId", "appId"),
        "audioGetBytes" to setOf("requestId", "audioId", "appId"),
        "audioList" to setOf("requestId", "filter"),
        "audioDelete" to setOf("requestId", "audioId", "appId"),
        "audioRegisterResult" to setOf("requestId", "asset", "errorMessage"),
        "audioAttachTranscriptResult" to setOf("requestId", "transcript", "errorMessage"),
        "audioGetResult" to setOf("requestId", "asset", "errorMessage"),
        "audioBytesResult" to setOf("requestId", "audioBase64", "mimeType", "durationMs", "errorMessage"),
        "audioListResult" to setOf("requestId", "list", "errorMessage"),
        "audioDeleteResult" to setOf("requestId", "deleted", "errorMessage"),
    )

    fun validateTopLevel(type: String, obj: JsonObject) {
        val payloadKeys = allowedPayloadKeys[type] ?: error("unknown frame type $type")
        val allowed = topLevelKeys + payloadKeys
        val extra = obj.keys.firstOrNull { it !in allowed }
        require(extra == null) { "unexpected field for $type: $extra" }
    }
}

File: packages/ClawixCore/Sources/ClawixCore/BridgeFrameValidation.swift
import Foundation

enum BridgeFrameValidation {
    static let topLevelKeys: Set<String> = ["schemaVersion", "type"]

    static func validate(_ data: Data, maxBytes: Int) throws {
        guard data.count <= maxBytes else {
            throw BridgeDecodingError.oversizedFrame(actualBytes: data.count, maxBytes: maxBytes)
        }

        let raw: Any
        do {
            raw = try JSONSerialization.jsonObject(with: data)
        } catch {
            throw BridgeDecodingError.invalidJSON(normalize(error))
        }

        guard let object = raw as? [String: Any] else {
            throw BridgeDecodingError.nonObjectFrame
        }

        guard object.keys.contains("schemaVersion") else {
            throw BridgeDecodingError.missingField("schemaVersion")
        }
        guard let schemaVersion = object["schemaVersion"] as? Int else {
            throw BridgeDecodingError.invalidField("schemaVersion")
        }
        guard schemaVersion == bridgeSchemaVersion else {
            throw BridgeDecodingError.unknownSchemaVersion(schemaVersion)
        }

        guard object.keys.contains("type") else {
            throw BridgeDecodingError.missingField("type")
        }
        guard let type = object["type"] as? String, !type.isEmpty else {
            throw BridgeDecodingError.invalidField("type")
        }
        guard let payloadKeys = allowedPayloadKeys(for: type) else {
            throw BridgeDecodingError.unknownType(type)
        }

        let allowed = topLevelKeys.union(payloadKeys)
        if let extra = object.keys.first(where: { !allowed.contains($0) }) {
            throw BridgeDecodingError.unknownField(type: type, field: extra)
        }
    }

    static func wrapDecodeError(_ error: Error, type: String) -> BridgeDecodingError {
        if let bridgeError = error as? BridgeDecodingError {
            return bridgeError
        }
        return .invalidPayload(type: type, message: normalize(error))
    }

    static func frameType(in data: Data) -> String? {
        guard let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return nil
        }
        return object["type"] as? String
    }

    private static func allowedPayloadKeys(for type: String) -> Set<String>? {
        switch type {
        case "auth":
            return keys(.token, .deviceName, .clientKind, .clientId, .installationId, .deviceId)
        case "listSessions", "pairingStart", "listProjects", "requestRateLimits", "requestClawJSServiceStatuses":
            return []
        case "openSession":
            return keys(.sessionId, .limit)
        case "loadOlderMessages":
            return keys(.sessionId, .beforeMessageId, .limit)
        case "sendMessage", "newSession":
            return keys(.sessionId, .text, .attachments)
        case "interruptTurn", "archiveSession", "unarchiveSession", "pinSession", "unpinSession":
            return keys(.sessionId)
        case "authOk":
            return keys(.hostDisplayName)
        case "authFailed":
            return keys(.reason)
        case "versionMismatch":
            return keys(.serverVersion)
        case "sessionsSnapshot":
            return keys(.sessions)
        case "sessionUpdated":
            return keys(.session)
        case "messagesSnapshot", "messagesPage":
            return keys(.sessionId, .messages, .hasMore)
        case "messageAppended":
            return keys(.sessionId, .message)
        case "messageStreaming":
            return keys(.sessionId, .messageId, .content, .reasoningText, .finished)
        case "errorEvent":
            return keys(.code, .message)
        case "editPrompt":
            return keys(.sessionId, .messageId, .text)
        case "renameSession":
            return keys(.sessionId, .title)
        case "readFile":
            return keys(.path)
        case "pairingPayload":
            return keys(.qrJson, .token, .shortCode)
        case "projectsSnapshot":
            return keys(.projects)
        case "fileSnapshot":
            return keys(.path, .content, .isMarkdown, .error)
        case "transcribeAudio":
            return keys(.requestId, .audioBase64, .mimeType, .language)
        case "transcriptionResult":
            return keys(.requestId, .text, .errorMessage)
        case "requestAudio":
            return keys(.audioId)
        case "audioSnapshot":
            return keys(.audioId, .audioBase64, .mimeType, .errorMessage)
        case "requestGeneratedImage":
            return keys(.path)
        case "generatedImageSnapshot":
            return keys(.path, .dataBase64, .mimeType, .errorMessage)
        case "requestRolloutAttachment":
            return keys(.attachmentId)
        case "rolloutAttachmentSnapshot":
            return keys(.attachmentId, .dataBase64, .mimeType, .errorMessage)
        case "bridgeState":
            return keys(.state, .chatCount, .message)
        case "rateLimitsSnapshot", "rateLimitsUpdated":
            return keys(.rateLimits, .rateLimitsByLimitId)
        case "clawJSServiceStatusesSnapshot":
            return keys(.services)
        case "clawJSServiceStatusUpdated":
            return keys(.service)
        case "audioRegister":
            return keys(.requestId, .request)
        case "audioAttachTranscript":
            return keys(.requestId, .audioId, .transcript)
        case "audioGet", "audioGetBytes", "audioDelete":
            return keys(.requestId, .audioId, .appId)
        case "audioList":
            return keys(.requestId, .filter)
        case "audioRegisterResult":
            return keys(.requestId, .asset, .errorMessage)
        case "audioAttachTranscriptResult":
            return keys(.requestId, .transcript, .errorMessage)
        case "audioGetResult":
            return keys(.requestId, .asset, .errorMessage)
        case "audioBytesResult":
            return keys(.requestId, .audioBase64, .mimeType, .durationMs, .errorMessage)
        case "audioListResult":
            return keys(.requestId, .list, .errorMessage)
        case "audioDeleteResult":
            return keys(.requestId, .deleted, .errorMessage)
        default:
            return nil
        }
    }

    private static func keys(_ keys: BridgePayloadKeys...) -> Set<String> {
        Set(keys.map(\.rawValue))
    }

    private static func normalize(_ error: Error) -> String {
        if let decodingError = error as? DecodingError {
            return normalize(decodingError)
        }
        let nsError = error as NSError
        if !nsError.localizedDescription.isEmpty {
            return nsError.localizedDescription
        }
        return String(describing: error)
    }

    private static func normalize(_ error: DecodingError) -> String {
        switch error {
        case .keyNotFound(let key, _):
            return "missing field \(key.stringValue)"
        case .typeMismatch(_, let context), .valueNotFound(_, let context), .dataCorrupted(let context):
            let path = context.codingPath.map(\.stringValue).joined(separator: ".")
            return path.isEmpty ? context.debugDescription : "\(path): \(context.debugDescription)"
        @unknown default:
            return String(describing: error)
        }
    }
}
